src/workflows/transport/pika_transport.py

```diff
@@ -1140,25 +1142,25 @@ def nack(
 
         # Check if channel is in tx mode
         transaction = self._transaction_on_channel.get(channel)
-        if transaction == transaction_id:
-            # Matching transaction IDs - perfect
-            self._channel_has_active_tx[channel] |= transaction is not None
-            self._connection.add_callback_threadsafe(
-                lambda: channel.basic_nack(
-                    delivery_tag, multiple=multiple, requeue=requeue
-                )
+        if transaction_id != transaction:
+            raise workflows.Error(
+                "Transaction state mismatch. "
+                f"Call assumes transaction {transaction_id}, channel has transaction {transaction}"
             )
-        elif transaction_id is None and not self._channel_has_active_tx[channel]:
+        if transaction_id is None and not self._channel_has_active_tx.get(channel):
 
             def _nack_callback():
                 channel.basic_nack(delivery_tag, multiple=multiple, requeue=requeue)
-                channel.tx_commit()
+                if self._channel_is_transactional.get(channel):
+                    channel.tx_commit()
 
             self._connection.add_callback_threadsafe(_nack_callback)
         else:
-            raise workflows.Error(
-                "Transaction state mismatch. "
-                f"Call assumes transaction {transaction_id}, channel has transaction {transaction}"
+            # Matching transaction IDs - perfect
+            # We are in a transaction so make a note that it is now being used
+            self._channel_has_active_tx[channel] = True
+            self._connection.add_callback_threadsafe(
+                lambda: channel.basic_nack(delivery_tag, multiple=multiple)
```

Comment:
What's the difference between above using `requeue=requeue` and this call without?

Reply:
This should be fixed by https://github.com/DiamondLightSource/python-workflows/pull/116/commits/baddff2bc55b2003f979a6a879d6e35730743b34 which adds an explicit test to cover this use case.

Reply:
Thanks. Looking good now!